### ISSUE TITLE
Server crashes with "Stack Overflow" when going into a specific mining area

### ISSUE THREAD

**Issue by D3nnis3n** [StrangeLoopGames/EcoIssues#18940]
Savegame: https://1drv.ms/u/s!AulmeOXOwLh2gbFrlIulOhRHiUsryw?e=ms5AIK
Logs:
[log_201026084714.log](https://github.com/StrangeLoopGames/EcoIssues/files/5445451/log_201026084714.log)
[ServerCrash_AggregateException_10241324.ecodmp.txt](https://github.com/StrangeLoopGames/EcoIssues/files/5445447/ServerCrash_AggregateException_10241324.ecodmp.txt)

Game is modded.

**Comment by mirasrael** [comment]
I investigates this issue. Seems happened thing I afraid in https://github.com/StrangeLoopGames/Eco/pull/4315
We have these stack traces:
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Eco.Simulation.RouteProbing.RouteRegions.GetRegion(WorldPosition3i pos, Int32 searchRange)
   at Eco.Mods.Organisms.Behaviors.FindAndEatPlantsBehavior.GetFoodSources(Animal agent)
   at Eco.Mods.Organisms.Behaviors.FindAndEatSomethingBehavior.FindAndEatSomething(Animal agent, Behavior`1 noNearbyFoodBehavior, Func`1 getFoodSources, Single hungerRestoredPerEat)+MoveNext()
   at Eco.Simulation.Agents.AI.BehaviorComplex`1.Do(T context)
   at Eco.Simulation.Agents.AI.BehaviorSelector`1.Do(T context)
   at Eco.Simulation.Agents.AI.Brain.TickBehaviors(Animal animal)
   at Eco.Simulation.Agents.AI.Brain.Tick(Animal animal)
   at Eco.Simulation.Agents.Animal.Tick()
   at Eco.Simulation.Simulation.TickAll()
   at Eco.Simulation.EcoSim.<DoTick>b__41_0()
   at Eco.Core.Plugins.TickTimeUtil.TimeSubprocess(Action func)
   at Eco.Simulation.EcoSim.DoTick(TickSample tick)
   at Eco.Simulation.EcoSim.DoWork(CancellationToken token)
   at Eco.Core.Utils.Threading.RepeatableWorkerBase.DoWork(CancellationToken token)
```
and
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Eco.Simulation.RouteProbing.RouteCacheData.SetRegionID(WorldPosition3i pos, Int16 regionID)
```

They both uses `TryGetWalkable` and then uses value as it isn't `null` (what is expected for TryGetValue):
```csharp
internal bool TryGetWalkable(WorldPosition3i groundPos, out T value)            
{ 
    return this.grid.TryGetValue(groundPos, out value);
}
```
but because we are using `grid` Dictionary (which is not thread-safe) in multi-threading environment for `TryGetValue` implementation it is possible to get wrong result:
```csharp
public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)
        {
            int i = FindEntry(key);
            if (i >= 0)
            {
                value = _entries![i].value;
                return true;
            }
            value = default!;
            return false;
        }
```
`null` value in this case.
Let's take a scenario when we remove item in one thread T1 and reading it in another thread T2:
T1: TryGetValue(key),             int i = FindEntry(key); (_entries![i].value != null)
T2: Remove(key), _entries[i] = null (_entries![i].value become null)
T1: value = _entries![i].value; (which already `null`)
So because of TryGetValue operation isn't atomic we have `null` where it shouldn't be and have a crash.

**Comment by D3nnis3n** [comment]
As this is confirmed to be on our side despite being modded, set high priority due to being a crash.

**Comment by mirasrael** [comment]
Actually I checked code and it shouldn't be this case, because value in this example is `struct` which never null.

**Comment by mirasrael** [comment]
So it seems it may only happen when `RouteManager.Obj.Grid` is null... which should never be `null` from code logic.

**Comment by mirasrael** [comment]
Also log doesn't seems relevant to topic subject. It says StackOverflowException, but in log I can see only Grid related error.

**Comment by mirasrael** [comment]
We need player name and exact position. If it happens every time and what we need to do to reproduce. Also we need information if this issue exists when mods disabled.

**Comment by mirasrael** [comment]
Fixed the issue. It happened because there was big unprocessed walkable region (in the mine). When new path point was added to this region it tried to recursively process all connected waypoints, but because of huge area it end up with too deep call stack. It usually shouldn't happen, because in normal situation you can't add lot of unprocessed path points at once, but it seems for some reason Route Data was corrupted (which also may be related to other errors in the log) and either reset (after restart) or not fully saved which caused this issue with erased routing data for large area. I changed logic to not use recursion and use optimized logic in such and similar cases.

**Comment by mirasrael** [comment]
Fix provided for 9.1, but if you don't want to wait then hot-fixed 0.9.0.5 version available here: https://drive.google.com/file/d/1LlzokLbMoLmN9jMeXawbhjSEOFaAAtLW/view?usp=sharing
